### PR TITLE
d2ir: Single Level Ampersand Filters

### PR DESCRIPTION
**Pull request by nhooyr**
Depends on #1479 being merged.

I didn't implement recursive filters in this PR. i.e you can't do something like:

```d2
*: {
  &style: {
    fill: red
    multiple: true
  }
  label: red and multiple
}
```

You have to do the following instead:
```d2
*: {
  style: {
    &fill: red
    &multiple: true
  }
  label: red and multiple
}
```

### CODE REVIEW

**Comment by nhooyr** [comment]
In fact I don't think we want to have composite filters anyhow. I think single field filters are easier to read and much easier to implement. For example for composite filters we'd have to error if the filter itself had ampersand filters because that wouldn't make sense.

File: d2ir/compile.go [hunk 1]
```diff
@@ -396,20 +410,87 @@ func (c *compiler) compileKey(refctx *RefContext) {
...
+	for _, f := range fa {
+		c._compileField(f, refctx)
+	}
+}
+
+func (c *compiler) ampersandFilter(refctx *RefContext) bool {
+	if !refctx.Key.Ampersand {
+		return true
+	}
+	if len(refctx.Key.Edges) > 0 {
```
Comment (alixander) [inline d2ir/compile.go:432]: is ampersand not functional for edges?

```d2
fast: {
  a
  far
}

task: {
  a
}

rats

* -> rats: {
  &label: task
}
```

Screen Shot 2023-07-30 at 11 42 05 AM

File: d2ir/compile.go [hunk 2]
```diff
@@ -396,20 +410,87 @@ func (c *compiler) compileKey(refctx *RefContext) {
...
+	return true
+}
+
+func (c *compiler) _ampersandFilter(f *Field, refctx *RefContext) bool {
+	f2 := ParentMap(f).Map().GetField(refctx.Key.Key.IDA()...)
+	if f2 == nil {
+		return false
+	}
+	if refctx.Key.Value.ScalarBox().Unbox() == nil {
+		c.errorf(refctx.Key, "ampersand filters cannot be composites")
```
Comment (alixander) [inline d2ir/compile.go:459]: i think we should call them "glob filters" user-facing.

**Review by alixander — CHANGES_REQUESTED** [review]

File: d2ir/compile.go (hunk 1 above)
Comment (nhooyr) [reply d2ir/compile.go:432]: Ah I forgot will fix.

Comment (nhooyr) [reply d2ir/compile.go:432]: Oh no so it does work on edge maps. But you can't use ampersands to filter maps with edges. E.g. you can't say apply this to a map with an edge from y -> z.

Comment (nhooyr) [reply d2ir/compile.go:432]: Nvm there was a bug. fixed. See test.

Comment (alixander) [reply d2ir/compile.go:438]: i don't understand. the test doesn't test this case. after latest pull, i now get `glob filters cannot be used outside globs` with the same code.

**Review by alixander — APPROVED** [review]

File: d2ir/compile.go (hunk 1 above)
Comment (nhooyr) [reply d2ir/compile.go:438]: I'm not sure what you're trying to do in your test there.

**Pull request merged**